webpush/encrypt.go

```diff
@@ -241,11 +292,42 @@ func newInfo(infoType string, context []byte) []byte {
 	info = append(info, []byte("Content-Encoding: ")...)
 	info = append(info, []byte(infoType)...)
 	info = append(info, 0)
-	info = append(info, []byte("P-256")...)
-	info = append(info, context...)
+	// For aes128gcm ctx is not needed.
+	if len(context) > 0 {
+		info = append(info, []byte("P-256")...)
+		info = append(info, context...)
+	}
 	return info
 }
 
+// Returns a keying material. See sections of
+// https://tools.ietf.org/html/draft-ietf-webpush-encryption-07#section-3.3
+func newKeyInfo(userAgentPublicKey, serverPublicKey []byte) []byte {
+	var info []byte
+	info = append(info, keyInfoPrefix...)
+	info = append(info, userAgentPublicKey...)
+	info = append(info, serverPublicKey...)
+	return info
+}
+
+// Returns an 86 octet header. See section 2.1 of:
+// https://tools.ietf.org/html/draft-ietf-httpbis-encryption-encoding-06#section-2.1
+func appendHeader(salt, serverPublicKey, ciphertext []byte) []byte {
+	var result []byte
+	// A push service is not required to support more than 4096 octets of
+	// payload body so the record size can be at most 4096.
+	cplen := uint32(4096)
```

Comment:
The intermediary values in the spec seems to enforce, that's why so did I: https://tools.ietf.org/html/draft-ietf-webpush-encryption-08#appendix-A

And I believe the RFC tests also depend on it, since that's the way it's used in the spec.

For cplen: I think I just copied from the existing version but it's referring to client public key length probably=) Will change it to plen for payload length